### PR TITLE
Add SortArrayInClass benchmark.

### PR DESCRIPTION
**Pull request by atrick**
Add SortArrayInClass benchmark.
    
    This currently copies the array each time it swaps elements. This
    makes it 1500x slower than it should be to sort the array. The
    benchmark now runs in 15ms but should be around 10us when fully
    optimized.
    
    This algorithm is an interesting optimization problem involving array
    optimization, uniqueness, bounds checks, and exclusivity. But the
    general first order problem is how to modify a CoW data structure
    that's stored in a class property. As it stands, the property getter
    retains the class property around the modify accesses that checks
    uniqueness.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci benchmark

File: benchmark/single-source/SortArrayInClass.swift [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  private func _swap(i: Int, j: Int) {
+    let t = array[i]
+    array[i] = array[j]
+    array[j] = t
+  }
+
+  private func _quicksort(left: Int, right: Int) {
+
+    if left < right {
+      let pivot = array[left &+ ((right &- left) / 2)]
```
Comment (compnerd) [inline benchmark/single-source/SortArrayInClass.swift]: Might be even more interesting to remove this attempt to reduce some of the instructions and use `+`, `-` instead.

**Review by compnerd — APPROVED** [review]
Thanks for the help in getting this added.

**Comment by jrose-apple** [comment]
Isn't this correct behavior, since there could be a subclass that overrides `array`?

**Comment by atrick** [comment]
@swift-ci smoke test.

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@jrose-apple It's true that if the class were `open` *and* the variable was `public`, or if we didn't use `wmo`, then we wouldn't be able to inline the array accessor. But the benchmark class is internal and it's built with `-O -wmo`, so the array accessor does get inlined. That's beside the point though because I'm questioning whether the array should be copied even at `-Onone`.

From any normal programmers's point of view this code:

    let t = array[i]
    array[i] = array[j]
    array[j] = t

Should be equivalent to:

    let t = array[i]
    let u = array[j]
    array[i] = u
    array[j] = t

The fact that we force an array copy here seems like an unintended artifact of SILGen to me rather than fundamental lanuage semantics. For some reason the scope of the assignment's RHS extends beyond the modification rather than ending after the RHS is fully evaluated. @rjmccall will remember why and know if there's a good reason we can't fix it.

**Comment by jrose-apple** [comment]
Oh, yeah, sure, the problem is a real problem. My question was whether the benchmark made any sense, and you're right, it does.

**Pull request merged**